packages/react/src/utils/useScrollLock.ts

```diff
@@ -137,66 +140,92 @@ function preventScrollStandard(referenceElement: Element | null) {
+      return;
+    }
+
+    const isOverflowHiddenLock = isIOS() || (isFirefox() && !hasInsetScrollbars(referenceElement));
+
+    // Firefox on macOS with overlay scrollbars uses a basic scroll lock that doesn't
+    // need the inset scrollbars handling to prevent overlay scrollbars from appearing
+    // on scroll containers briefly whenever the lock is enabled.
+    // On iOS, scroll locking does not work if the navbar is collapsed. Due to numerous
+    // side effects and bugs that arise on iOS, it must be researched extensively before
+    // being enabled to ensure it doesn't cause the following issues:
+    // - Textboxes must scroll into view when focused, nor cause a glitchy scroll animation.
+    // - The navbar must not force itself into view and cause layout shift.
+    // - Scroll containers must not flicker upon closing a popup when it has an exit animation.
+    this.restore = isOverflowHiddenLock
+      ? preventScrollBasic(referenceElement)
+      : preventScrollStandard(referenceElement);
+  }
+}
+
+const SCROLL_LOCKER = new ScrollLocker();
+
 /**
  * Locks the scroll of the document when enabled.
  *
  * @param enabled - Whether to enable the scroll lock.
  */
 export function useScrollLock(params: {
-  enabled?: boolean;
+  enabled: boolean;
   mounted: boolean;
   open: boolean;
   referenceElement?: Element | null;
 }) {
   const { enabled = true, mounted, open, referenceElement = null } = params;
 
-  const isOverflowHiddenLock = React.useMemo(
-    () => enabled && (isIOS() || (isFirefox() && !hasInsetScrollbars(referenceElement))),
-    [enabled, referenceElement],
-  );
-
-  useModernLayoutEffect(() => {
-    // https://github.com/mui/base-ui/issues/1135
-    if (mounted && !open && isWebKit()) {
-      const doc = ownerDocument(referenceElement);
-      const originalUserSelect = doc.body.style.userSelect;
-      const originalWebkitUserSelect = doc.body.style.webkitUserSelect;
-      doc.body.style.userSelect = 'none';
-      doc.body.style.webkitUserSelect = 'none';
-      return () => {
-        doc.body.style.userSelect = originalUserSelect;
-        doc.body.style.webkitUserSelect = originalWebkitUserSelect;
-      };
-    }
-    return undefined;
-  }, [mounted, open, referenceElement]);
+  // https://github.com/mui/base-ui/issues/1135
+  if (isWebKit()) {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    useModernLayoutEffect(() => {
+      if (mounted && !open) {
+        const doc = ownerDocument(referenceElement);
+        const originalUserSelect = doc.body.style.userSelect;
+        const originalWebkitUserSelect = doc.body.style.webkitUserSelect;
+        doc.body.style.userSelect = 'none';
+        doc.body.style.webkitUserSelect = 'none';
+        return () => {
+          doc.body.style.userSelect = originalUserSelect;
+          doc.body.style.webkitUserSelect = originalWebkitUserSelect;
+        };
+      }
+      return undefined;
+    }, [mounted, open, referenceElement]);
+  }
```

Comment:
Not sure if this is worth it? We should try to be React Compiler-compliant as much as possible and avoid disabling lint rules. It also seems a bit risky based on how `isWebKit` is calculated and could lead to a crash in some scenarios (I thought server pre-rendering -> client hydration in Safari might be an issue, but can't reproduce)

Reply:
> It also seems a bit risky based on how isWebKit is calculated

Making those functions into constants is on my todo list, but I didn't want to busy this PR with those changes.

> We should try to be React Compiler-compliant as much as possible and avoid disabling lint rules

I don't insist much on keeping this optimization, but out of curiosity are you aware of any upcoming issues with that?

Reply:
Last time I tested it (awhile ago, last year) it bailed out of optimizing a hook or component if a lint rule was disabled